tests/qemu-tests/conftest.py

```diff
@@ -0,0 +1,109 @@
+"""
+This file should consist of global test fixtures.
+"""
+
+from __future__ import annotations
+
+import os
+import subprocess
+import sys
+
+import gdb
+import pytest
+from pwn import context
+from pwn import make_elf_from_assembly
+
+_start_binary_called = False
+
+QEMU_PORT = os.environ.get("QEMU_PORT")
+
+
+@pytest.fixture
+def qemu_assembly_run():
+    """
+    Returns function that launches given binary with 'starti' command
+
+    The `path` is returned from `make_elf_from_assembly` (provided by pwntools)
+    """
+
+    qemu: subprocess.Popen = None
+
+    if QEMU_PORT is None:
+        print("'QEMU_PORT' environment variable not set")
+        sys.exit(1)
+
+    def _start_binary(asm: str, arch: str, *args):
+        nonlocal qemu
+
+        context.arch = arch
+        binary_tmp_path = make_elf_from_assembly(asm)
+
+        qemu = subprocess.Popen(
+            [
+                f"qemu-{arch}",
+                "-g",
+                f"{QEMU_PORT}",
+                f"{binary_tmp_path}",
+            ]
+        )
+
+        gdb.execute(f"target remote :{QEMU_PORT}")
+        gdb.execute("set exception-verbose on")
+
+        global _start_binary_called
+        # if _start_binary_called:
+        #     raise Exception('Starting more than one binary is not supported in pwndbg tests.')
+
+        _start_binary_called = True
+
+    yield _start_binary
+
+    qemu.terminate()
+
+    # Ensure qemu is stopped
+
+
+@pytest.fixture
+def qemu_start_binary():
+    """
+    Returns function that launches given binary with 'starti' command
+
+    The `path` is returned from `make_elf_from_assembly` (provided by pwntools)
+    """
+
+    qemu: subprocess.Popen = None
+
+    if QEMU_PORT is None:
+        print("'QEMU_PORT' environment variable not set")
+        sys.exit(1)
+
+    def _start_binary(path: str, arch: str, *args):
+        nonlocal qemu
+
+        qemu = subprocess.Popen(
+            [
+                "setarch",
+                "-R",
+                f"qemu-{arch}",
+                "-L",
+                f"/usr/{arch}-linux-gnu/",
+                "-g",
+                f"{QEMU_PORT}",
+                f"{path}",
+            ]
+        )
+
+        gdb.execute(f"target remote :{QEMU_PORT}")
+        gdb.execute("set exception-verbose on")
+
+        global _start_binary_called
+        # if _start_binary_called:
+        #     raise Exception('Starting more than one binary is not supported in pwndbg tests.')
+
+        _start_binary_called = True
+
+    yield _start_binary
+
+    qemu.terminate()
```

Comment:
Shall we use `qemu.kill()` instead?

Per the [`Popen.terminate()`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.terminate) and `kill` docs, one sends `SIGTERM` and another sends `SIGKILL`.

Per `man 7 signal`:
```
The signals SIGKILL and SIGSTOP cannot be caught, blocked, or ignored.
```

So if we want to have 100% reliability on _plz kill that qemu process_ then we should probably use `.kill()` here :).